jsonic3_commit.py

```diff
@@ -0,0 +1,80 @@
+        },
+    .
+    .
+    .
+    ]
+    etc
+In addition to an array of Question objects, each Section incorporates its title and a 'last updated' date value. For the Python intermediary stage (between CSV stream  and JSON) we will create a hierarchy of nested dictionaries and lists equivalent to the JSON objects and arrays.
+"""
+
+import requests
+import json
+import csv
+import datetime
+from typing import Iterable, Dict, List, Any
+
+
+# function to grab data from a google sheet + return as reader object
+def google_sheet_csv_data(sheet: str, gid: str) -> Iterable[List[str]]:
+    url = f'https://docs.google.com/spreadsheets/d/{sheet}/export'
+    response = requests.get(url, params={
+        'format': 'csv',
+        'id': sheet,
+        'gid': gid
+    })
+    response.raise_for_status()
+    return csv.reader(response.iter_lines(decode_unicode=True))
+
+
+# Pass faq Content sheet id and gid to function
+_reader = google_sheet_csv_data('1_wBXS62S5oBQrwetGc8_-dFvDjEmNqzqHwUeP-DzkYs',
+ '1318925039')
+
+# Date, to be updated each time the program runs, in UTC
+date = datetime.datetime.utcnow().strftime('%Y-%m-%d')
+
+# Create the main FAQ dictionary
+# Create a list to contain our Section objects
+# Place the Sections list inside the main FAQ dictionary
+faq_dict = {}
+sections_list: List[Dict] = []
+faq_dict['faqItems'] = sections_list
+
+# Ensure only first two columns in each row are used
+# Now arrange the Sections and Questions in a JSON-friendly way
+for index, row in enumerate(_reader):
+    rowtype, rowval, *_rest = row
+    if rowtype == 'Category':   # We don't need this row
+        pass
+    elif rowtype == 'Section Head':
+        section: Dict[str, Any] = {}   # Create a dictionary for a Section
+        sections_list.append(section)   # Add new Section to Sections list
+        section['title'] = rowval   # Give the Section its title
+        section['lastUpdatedAt'] = date   # a 'last updated' value
+        questions_list: List[Dict] = []   # Create list to contain questions
+        section['qa'] = questions_list   # add new Questions list to its Section
+    elif rowtype == 'Q':
+        question = {}   # Create new Question dictionary
+        question['q'] = rowval   # the Question's title
+        questions_list.append(question)   # Append Question to given Section
+    elif rowtype == 'A':   # Now the same for Answers & Links
+        question['a'] = rowval
+    elif rowtype == 'link':
+        question['url'] = rowval
+    elif rowtype == '':   # Some safeguards
+        pass
+    elif rowval and not rowtype:
+        raise ValueError(f'row {index} has a value but no description!')
+    else:
+        raise ValueError(f'Unknown row header: "{rowtype}"')
+
+# Create formatted json file
+with open('faq.json', 'w') as f:   # Create JSON output file
+    json.dump(faq_dict, f, indent=2)   # Put FAQ Content in output file as JSON
```

Comment:
I generally find scripts like this to be much more useful and flexible if they print their result to STDOUT (i.e. `print(result)` or `sys.stdout.write(result)`) rather than writing to a predetermined file. That way you can put the results wherever you want when running it:

```sh
$ python jsonic3_commit.py > path/where/i/want/faq.json
```

When you’ve got multiple outputs or more complex output, you can provide a `-o` or `--output` option for specifying where to write the output, like we do with `scraper_data.py` or `scraper_news.py`:
https://github.com/sfbrigade/data-covid19-sfbayarea/blob/06342ce29cf95b68c115cf0660ec904489e7f5b7/scraper_news.py#L18-L19
https://github.com/sfbrigade/data-covid19-sfbayarea/blob/06342ce29cf95b68c115cf0660ec904489e7f5b7/scraper_news.py#L38-L44

But that’s not a hard requirement. I just think it tends to work a lot better.

Reply:
Yeah, this is great - thanks v much for the steer...enjoyed reading all the docs and learning this and will def action.